pkg/net/net.go

```diff
@@ -22,10 +24,65 @@ type Message interface {
 
 	Payload() interface{}
 
-	Type() string
+	Type() MessageType
 	Seqno() uint64
 }
 
+// TypedPayload is a constraint type representing an abstract network payload
+// that describes its own type. Despite the Message interface extends TypedPayload,
+// there is no guarantee that Message.Payload() does the same. The TypedPayload
+// should be used in the context of Message.Payload() only in cases where the
+// underlying Message.Payload() type actually implements TypedPayload.
```

Comment:
I am staring at this PR for almost 3 hours now and I am quite convinced, `pkg/net` is not the right place for this code.

Just like the documentation for `TypedPayload` says, some payloads have `Type()` but this is not generally guaranteed. As noted in the other comment to this PR, the fact payloads have `Type()` is our simplification and not something `net.BroadcastChannel` interface expects.

Next, we deduplicate messages based on a condition that is out of the scope of `pkg/net` and comes from the protcol. This is not retransmission deduplication or floodsub deduplication the `pkg/net` should take care of. This creature does not come from `pkg/net` but from the protocol and I am afraid `func (tp TypedPayloads[T]) Deduplicate(keyFn func(T) string) TypedPayloads[T]` living in `pkg/net` would be a common point of confusion of what this deduplication is actually about.

Rephrasing what I think is a problem here: we are deduplicating *some* `net.Message`s (not all of them are "deduplicatable") based on a protocol-level requirements (abstracted out, but still coming from the protocol - this is not a network-level deduplication, it is a protocol-specific deduplication).

Naturally, the question arises: if not `pkg/net` then where and how? 

My first thought was that this is something `pkg/protocol` should take care of. We are deduplicating received messages of the given type based on some protocol rules supplied. So this could be a common code in `pkg/protocol/message`, for example.

On the other hand, "deduplicating messages of the given type" is primarily a problem of an asynchronous state machine. And the async state machine is the reason we have this problem in the first place.

The given state of a synchronous state machine accepts only one type of message and can do the deduplication at the moment the message is received. In case of the async state machine, we are accepting all types of messages, also those from the future, and the deduplication can be done only when we finally reach the state understanding the given type of messages and knowing how to deduplicate them.

So this could also be a logic in `BaseAsyncState`.

I am personally leaning towards this option. It is the simplest one because I think we can avoid `TypedPayload`, `TypedPayloads`, and `TypedMessagesSupplier`. If we realize one day, sync state machine could also find this pattern useful, we can move it to `pkg/protocol/message` and play with typed abstractions. 

What do you think?

Reply:
See https://github.com/keep-network/keep-core/pull/3386/commits/96cb66c036b77ad4fbec927a680164afc0cb6174